system/jlib/jptree.cpp

```diff
@@ -7609,48 +7667,82 @@ static constexpr const char * currentVersion = "1.0";
+ *   Elements with a name attribute are matched by name.  If there is a match it is merged.  If there is no match it is added.
 */
 
-void mergeConfiguration(IPropertyTree & target, IPropertyTree & source)
+static bool checkInSequence(IPropertyTree & child, StringAttr &seqname, bool &first, bool &endprior)
+{
+    first = false;
+    endprior = false;
+    if (seqname.length() && streq(seqname, child.queryName()))
+        return true;
+    endprior = !seqname.isEmpty();
+    if (child.isArray(nullptr))
+    {
+        first=true;
+        seqname.set(child.queryName());
+        return true;
+    }
+    seqname.clear();
+    return false;
+}
+
+inline bool isScalarItem(IPropertyTree &child)
+{
+    if (child.hasChildren())
+        return false;
+    return child.getAttributeCount()==0;
+}
+
+static IPropertyTree *ensureMergeConfigTarget(IPropertyTree &target, const char *tag, const char *name, bool sequence)
+{
+    StringBuffer tempPath;
+    const char * path = (sequence) ? nullptr : tag;
+    if (name)
+    {
+        tempPath.clear().append(tag).append("[@name=\'").append(name).append("']");
+        path = tempPath;
+    }
+
+    IPropertyTree * match = (path) ? target.queryPropTree(path) : nullptr;
+    if (!match)
+    {
+        if (sequence)
+            match = target.addPropTreeArrayItem(tag, createPTree(tag));
+        else
+            match = target.addPropTree(tag);
+    }
+    return match;
+}
+
+void mergeConfiguration(IPropertyTree & target, IPropertyTree & source, const char *altNameAttribute)
 {
     Owned<IAttributeIterator> aiter = source.getAttributes();
     ForEach(*aiter)
         target.addProp(aiter->queryName(), aiter->queryValue());
 
-    StringBuffer tempPath;
+    StringAttr seqname;
     Owned<IPropertyTreeIterator> iter = source.getElements("*");
     ForEach(*iter)
     {
         IPropertyTree & child = iter->query();
         const char * tag = child.queryName();
         const char * name = child.queryProp("@name");
-        //Legacy support for old roxie configuration files that have repeated elements with no name tag
+        //Legacy support for old component configuration files that have repeated elements with no name tag but another unique id
         if (!name)
-            name = child.queryProp("@netAddress");
-        const char * path = tag;
-        if (name)
-        {
-            tempPath.clear().append(path).append("[@name=\'").append(name).append("']");
-            path = tempPath;
-        }
-        if (child.queryProp("@__remove__"))
-        {
-            target.removeProp(path);
-        }
-        else
-        {
-            IPropertyTree * match = target.queryPropTree(path);
-            if (!match)
-            {
-                match = target.addPropTree(tag);
-                if (name)
-                    match->setProp("@name", name);
-            }
-            mergeConfiguration(*match, child);
-        }
+            name = child.queryProp(altNameAttribute);
```

Comment:
@afishbeck - should this check if altNameAttribute is set and take some other default action?

Reply:
oops, yes.